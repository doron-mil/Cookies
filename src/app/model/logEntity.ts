export interface LogEntity {
  id: number;
  content: string;
}
