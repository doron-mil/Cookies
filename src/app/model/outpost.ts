export class Outpost {
  id: number;
  name: string;
  location: string;
  supplyMeter: number;
  logNumber: number;
}
