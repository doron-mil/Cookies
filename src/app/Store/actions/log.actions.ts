import {LogEntity} from '../../model/logEntity';
import {OUTPOST_FEATURE} from './outpost.actions';

export const LOG_FEATURE = '[LOGS]';
export const LOG_OFFLINE_FEATURE = '[LOGS-OFFLINE]';

// define action types
export const Add_LOG = `${LOG_FEATURE} ADD`;
export const ADD_LOG_SUCCESS = `${LOG_FEATURE} SUCCESS `;
export const ADD_MULTI_LOGS = `${LOG_FEATURE} ADD MULTI LOGS`;

export const ADD_LOG_OFFLINE_ACTION = `${LOG_OFFLINE_FEATURE} ADD_OFFLINE`;
export const ADD_LOG_OFFLINE_COMMIT_ACTION = `${LOG_OFFLINE_FEATURE} ADD_OFFLINE_COMMIT`;
export const ADD_LOG_OFFLINE_ROLLBACK_ACTION = `${LOG_OFFLINE_FEATURE} ADD_OFFLINE_ROLLBACK`;
export const ADD_LOG_OFFLINE_BULK_ACTION = `${LOG_OFFLINE_FEATURE} ADD_OFFLINE_BULK`;

// define action constructors

export const addLogActionOffline = (logEntity: LogEntity, url: string) => ({
  type: ADD_LOG_OFFLINE_ACTION,
  payload: logEntity,
  meta: {
    feature: LOG_OFFLINE_FEATURE,
    offline: {
      // the network action to execute:
      effect: {url: url, method: 'GET', body: null},
      // action to dispatch when effect succeeds:
      commit: {type: ADD_LOG_OFFLINE_COMMIT_ACTION, meta: {feature: LOG_OFFLINE_FEATURE, logEntity}},
      // action to dispatch if network action fails permanently:
      rollback: {type: ADD_LOG_OFFLINE_ROLLBACK_ACTION, meta: {feature: LOG_OFFLINE_FEATURE, logEntity}}
    }
  },
});

export const addBulkLogActionOffline = (logEntities: LogEntity[]) => ({
  type: ADD_LOG_OFFLINE_BULK_ACTION,
  meta: {feature: LOG_OFFLINE_FEATURE},
  payload: logEntities,
});


export const addLogAction = (logEntity: LogEntity) => ({
  type: Add_LOG,
  payload: logEntity,
  meta: {
    feature: LOG_FEATURE,
  },
});

export const addMultiLogs = (logEntities: LogEntity[]) => ({
  type: ADD_MULTI_LOGS,
  payload: logEntities,
  meta: {
    feature: LOG_FEATURE,
  },
});

export const addLogSuccess = () => ({
  type: ADD_LOG_SUCCESS,
  meta: {feature: LOG_FEATURE},
});




