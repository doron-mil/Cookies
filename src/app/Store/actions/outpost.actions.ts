import {Outpost} from '../../model/outpost';

export const OUTPOST_FEATURE = '[OUTPOST]';
export const OUTPOST_OFFLINE_FEATURE = '[OUTPOST-OFFLINE]';
export const OUTPOST_MASS_FEATURE = '[OUTPOST-MASS]';

// define action types
export const ADD_OUTPOST = `${OUTPOST_FEATURE} ADD`;
export const ADD_LOTS_OF_OUTPOSTS = `${OUTPOST_MASS_FEATURE} ADD_LOTS`;
export const ADD_OUTPOSTS_BULK = `${OUTPOST_MASS_FEATURE} ADD_BULK`;

export const ADD_OUTPOST_OFFLINE  = `${OUTPOST_FEATURE} ADD_OFFLINE`;
export const ADD_OUTPOST_OFFLINE_COMMIT = `${OUTPOST_FEATURE} ADD_OFFLINE_COMMIT`;
export const ADD_OUTPOST_OFFLINE_ROLLBACK = `${OUTPOST_FEATURE} ADD_OFFLINE_ROLLBACK`;

export const addOutpostAction = (outpost: Outpost) => ({
  type: ADD_OUTPOST,
  payload: outpost,
  meta: {
    feature: OUTPOST_FEATURE,
  }
});

export const addLotsOfOutpostsAction = () => ({
  type: ADD_LOTS_OF_OUTPOSTS,
  meta: {feature: OUTPOST_MASS_FEATURE}
});

export const addOutpostsBulkAction = (outpostArray: Outpost[]) => ({
  type: ADD_OUTPOSTS_BULK,
  payload: outpostArray,
  meta: {feature: OUTPOST_FEATURE}
});

