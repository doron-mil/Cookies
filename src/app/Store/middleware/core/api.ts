import {API_REQUEST, apiError, apiSuccess} from '../../actions/api.actions';

export const apiMiddleware = ({dispatch}) => (next) => (action) => {
  next(action);

  if (action.type.includes(API_REQUEST)) {
    const {body, url, method, feature} = action.meta;
    const {data} = action.data;

    fetch(url, {body, method})
      .then(response => response.json())
      .then(response => {
        // console.log('aaaaaaaaaaaa', response);
        dispatch(apiSuccess(response, feature, data));
      })
      .catch(error => dispatch(apiError(error, feature, data)));
  }
};

function type(obj) {
  return Object.prototype.toString.call(obj).slice(8, -1);
}
