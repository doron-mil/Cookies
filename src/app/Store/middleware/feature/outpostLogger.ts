import {
  ADD_LOTS_OF_OUTPOSTS,
  ADD_OUTPOST,
  addOutpostsBulkAction,
  OUTPOST_FEATURE,
  OUTPOST_MASS_FEATURE
} from '../../actions/outpost.actions';
import {API_ERROR, API_SUCCESS, apiRequest} from '../../actions/api.actions';
import {addLogAction} from '../../actions/log.actions';
import {Outpost} from '../../../model/outpost';

const API_TEST_URL = 'https://jsonplaceholder.typicode.com/posts/';
const API_TEST_PHOTOS_URL = 'https://jsonplaceholder.typicode.com/photos/';

interface PhotoInterface {
  albumId: number;
  id: number;
  title: string;
  url: string;
  thumbnailUrl: string;
}

let nextLogId: number = 1;

const getPayloadForAction = (name: string, title: string) => {
  title = title.slice(0, 30);
  const newID = nextLogId++;
  const logMsg = 'Added new outpost id = ' + newID + ', name = ' + name;
  const new_payload = {
    id: newID,
    content: logMsg,
    calculated: title,
  };
  return new_payload;
};


export const outpostMiddleware = ({getState, dispatch}) => (next) => (action) => {
  next(action);

  switch (action.type) {

    case ADD_OUTPOST:
      const outpost = action.payload;
      outpost.id = getState().outposts.outpostList.length;
      const url = API_TEST_URL + ((outpost.id % 100) + 1);
      next(apiRequest(null, 'GET', url, OUTPOST_FEATURE, outpost));
      break;

    case `${OUTPOST_FEATURE} ${API_SUCCESS}`:

      // console.log('3333333333', action);
      const payload = getPayloadForAction(action.data.name, action.payload.title);

      next(addLogAction(payload));

      break;

    case `${OUTPOST_FEATURE} ${API_ERROR}`:
      console.log('Failed the API');

      const err_payload = getPayloadForAction(action.data.name, 'NO CONNECTION');

      next(addLogAction(err_payload));

      break;


    case ADD_LOTS_OF_OUTPOSTS:
      next(apiRequest(null, 'GET', API_TEST_PHOTOS_URL, OUTPOST_MASS_FEATURE, null));
      break;

    case `${OUTPOST_MASS_FEATURE} ${API_SUCCESS}`:

      const photoesArray: Array<PhotoInterface> = action.payload;
      const outpostArray: Array<Outpost> = [];

      // console.log('aaaaaaaaaaa', photoesArray.length , getState().logs.logsArray );
      let counter = getState().outposts.outpostList.length;
      // console.log('bbbbbbbbbbbbbbb',counter );

      photoesArray.forEach((photo: PhotoInterface) => {
        const newOutpost = new Outpost();
        newOutpost.id = counter++;
        newOutpost.name = photo.title;
        newOutpost.location = photo.url + ' --- ' + photo.thumbnailUrl;
        newOutpost.supplyMeter = Math.floor(Math.random() * 101);
        newOutpost.logNumber = Math.floor(Math.random() * 501);

        outpostArray.push(newOutpost);
      });

      next(addOutpostsBulkAction(outpostArray));
      // dispatch(addLogsBulkAction(outpostArray));

      break;

    case `${OUTPOST_MASS_FEATURE} ${API_ERROR}`:
      console.log('Failed the API', action);
      break;

  }
};

function type(obj) {
  return Object.prototype.toString.call(obj).slice(8, -1);
}
