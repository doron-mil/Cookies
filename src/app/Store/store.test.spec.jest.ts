import 'jest';
import configureStore from 'redux-mock-store';
import {Add_LOG, addLogAction, LOG_FEATURE} from './actions/log.actions';
import {LogEntity} from '../model/logEntity';
import {logReducer} from './reducers/log.reducer';
import {INITIAL_APP_STATE, INITIAL_STATE} from './main.state';
import {outpostMiddleware} from './middleware/feature/outpostLogger';
import {addOutpostAction, OUTPOST_FEATURE} from './actions/outpost.actions';
import {Outpost} from '../model/outpost';
import {apiMiddleware} from './middleware/core/api'; // ES6 modules
import 'isomorphic-fetch';
import {apiError, apiRequest, apiSuccess} from './actions/api.actions';
import {promise} from 'selenium-webdriver';


describe('OutpostAddComponent', () => {

  beforeEach(() => {
  });

  beforeEach(() => {
  });

  const middlewares = [];
  const mockStore = configureStore(middlewares);

  const middlewares2 = [outpostMiddleware, apiMiddleware];
  const mockStore2 = configureStore(middlewares2);

  const delay = async ms => {
    await new Promise((resolve) => setTimeout(resolve, ms));
    return 1;
  };

  describe('#008 - API MIDDLEWARE - middleware itself using MOCKs', () => {
    const apiMiddlewareCreate = () => {
      const next = jest.fn();
      const store = {
        getState: jest.fn(() => ({outposts: INITIAL_APP_STATE, logs: INITIAL_STATE})),
        dispatch: jest.fn(),
      };
​
      const invoke = (action) => apiMiddleware(store)(next)(action);
      return {store, next, invoke};
    };

    it('#009 - API MIDDLEWARE CHECK - checking the middleware itself - apiError', async () => {
      // Dispatch the action
      const outpost: Outpost = {
        id: 1,
        name: 'OutpostTestName',
        location: 'TelAviv',
        supplyMeter: 269,
        logNumber: 1
      };

      const {store, next, invoke} = apiMiddlewareCreate();
      const action = apiRequest(null, 'GET', 'https://jsonplaceholder2.typicode.com/posts/a', OUTPOST_FEATURE, outpost);

      await invoke(action).then(() => {
        expect(next).toHaveBeenCalledTimes(1);
        expect(store.dispatch).toHaveBeenCalledTimes(1);
        expect(store.dispatch).toBeCalledWith(
          expect.objectContaining(apiError(expect.any(Object), OUTPOST_FEATURE, action.data)),
        );
      });
    });

    it('#008 - API MIDDLEWARE CHECK - checking the middleware itself', () => {
      // Dispatch the action
      const outpost: Outpost = {
        id: 1,
        name: 'OutpostTestName',
        location: 'TelAviv',
        supplyMeter: 269,
        logNumber: 1
      };

      const {store, next, invoke} = apiMiddlewareCreate();
      const action = apiRequest(null, 'GET', 'https://jsonplaceholder.typicode.com/posts/1', OUTPOST_FEATURE, outpost);

      // ******** Returning a promise
      return invoke(action).then(() => {
        expect(next).toHaveBeenCalledTimes(1);
        expect(store.dispatch).toHaveBeenCalledTimes(1);
        const response = {
          body: 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum' +
          '\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto',
          id: 1,
          title: 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit',
          userId: 1
        };
        expect(store.dispatch).toHaveBeenCalledWith(apiSuccess(response, OUTPOST_FEATURE, action.data));
      });
    });
  });


  describe('#007 - MIDDLEWARE - full scenario', () => {
    const store = mockStore2({outposts: INITIAL_APP_STATE, logs: INITIAL_STATE});

    beforeEach(() => {
      // Dispatch the action
      const outpost: Outpost = {
        id: 1,
        name: 'OutpostTestName',
        location: 'TelAviv',
        supplyMeter: 269,
        logNumber: 1
      };

      store.dispatch(addOutpostAction(outpost));
    });

    it('#007 - MIDDLEWARE CHECK - checking full - adding outpost will trigger 4 actions', async () => {
      await delay(2000);
      const actions = store.getActions();
      actions.forEach((action) => console.log('store.test.spec.jest - #006 - action ', action));
      expect(actions.length).toEqual(4);
      expect(actions).toMatchSnapshot();
    });
  });


  describe('#006 - MIDDLEWARE - middleware itself using MOCKs', () => {
    const outpostMiddlewareCreate = () => {
      const next = jest.fn();
      const store = {
        getState: jest.fn(() => ({outposts: INITIAL_APP_STATE, logs: INITIAL_STATE})),
        dispatch: jest.fn(),
      };
​
      const invoke = (action) => outpostMiddleware(store)(next)(action);
      return {store, next, invoke};
    };

    it('#006 - MIDDLEWARE CHECK - checking the middleware itself', () => {
      // Dispatch the action
      const outpost: Outpost = {
        id: 1,
        name: 'OutpostTestName',
        location: 'TelAviv',
        supplyMeter: 269,
        logNumber: 1
      };

      const {next, invoke} = outpostMiddlewareCreate();
      const action = addOutpostAction(outpost);
      invoke(action);
      expect(next).toHaveBeenCalledTimes(2);
      expect(next).toHaveBeenCalledWith(action);
      expect(next).toHaveBeenCalledWith(apiRequest(null, 'GET', 'https://jsonplaceholder.typicode.com/posts/1', OUTPOST_FEATURE, outpost));
    });
  });

  it('#005 - MIDDLEWARE CHECK - adding outpost will trigger 2 actions', () => {
    const store = mockStore2({outposts: INITIAL_APP_STATE});

    // Dispatch the action
    const outpost: Outpost = {
      id: 1,
      name: 'OutpostTestName',
      location: 'TelAviv',
      supplyMeter: 269,
      logNumber: 1
    };

    store.dispatch(addOutpostAction(outpost));

    // Test if your store dispatched the expected actions
    const actions = store.getActions();
    // const expectedPayload = {type: '[LOGS] ADD', payload: logEntity, meta: {feature: '[LOGS]'}};
    // actions.forEach((action) => console.log('store.test.spec.jest - #005 - action ', action));
    expect(actions.length).toEqual(2);
  });

  it('#004.1 - REDUCER CHECK - SNAPSHOT - using bad LogEntity ', () => {
    const logEntity = {
      type: Add_LOG,
      meta: {
        feature: LOG_FEATURE,
      },
    };
    const expected = {
      logsArray: [logEntity]
    };

    expect(logReducer(undefined, logEntity)).toMatchSnapshot();
  });

  it('#004 - REDUCER CHECK - SNAPSHOT - should return state with first log ', () => {
    const logEntity: LogEntity = {
      id: 1,
      content: 'testContent2',
      calculated: 'testCalculated'
    };
    const expected = {
      logsArray: [logEntity]
    };

    expect(logReducer(undefined, addLogAction(logEntity))).toMatchSnapshot();
  });

  it('#003 - REDUCER CHECK - should return state with first log', () => {
    const logEntity: LogEntity = {
      id: 1,
      content: 'testContent',
      calculated: 'testCalculated'
    };
    const expected = {
      logsArray: [logEntity]
    };

    expect(logReducer(undefined, addLogAction(logEntity))).toEqual(expected);
  });

  it('#002 - REDUCER CHECK - should return initial state', () => {
    expect(logReducer(undefined, {})).toEqual(INITIAL_STATE);
  });

  it('#001 - ACTION CHECK - should dispatch action', () => {
    // Initialize mockstore with empty state
    const initialState = {};
    const store = mockStore(initialState);

    // Dispatch the action
    const logEntity: LogEntity = {
      id: 1,
      content: 'testContent',
      calculated: 'testCalculated'
    };

    store.dispatch(addLogAction(logEntity));

    // Test if your store dispatched the expected actions
    const actions = store.getActions();
    const expectedPayload = {type: '[LOGS] ADD', payload: logEntity, meta: {feature: '[LOGS]'}};
    expect(actions).toEqual([expectedPayload]);
  });

  it('#000 - ACTION CHECK - checking action creation', () => {

    const logEntity: LogEntity = {
      id: 1,
      content: 'testContent',
      calculated: 'testCalculated'
    };

    const expectedAction = {type: '[LOGS] ADD', payload: logEntity, meta: {feature: '[LOGS]'}};
    expect(addLogAction(logEntity)).toEqual(expectedAction);
  });


})
;

