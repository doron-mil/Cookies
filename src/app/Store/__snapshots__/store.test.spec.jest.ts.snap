// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OutpostAddComponent #004 - REDUCER CHECK - SNAPSHOT - should return state with first log  1`] = `
Object {
  "logsArray": Array [
    Object {
      "calculated": "testCalculated",
      "content": "testContent2",
      "id": 1,
    },
  ],
}
`;

exports[`OutpostAddComponent #004.1 - REDUCER CHECK - SNAPSHOT - using bad LogEntity  1`] = `
Object {
  "logsArray": Array [
    undefined,
  ],
}
`;

exports[`OutpostAddComponent #007 - MIDDLEWARE - full scenario #007 - MIDDLEWARE CHECK - checking full - adding outpost will trigger 4 actions 1`] = `
Array [
  Object {
    "meta": Object {
      "feature": "[OUTPOST]",
    },
    "payload": Object {
      "id": 0,
      "location": "TelAviv",
      "logNumber": 1,
      "name": "OutpostTestName",
      "supplyMeter": 269,
    },
    "type": "[OUTPOST] ADD",
  },
  Object {
    "data": Object {
      "id": 0,
      "location": "TelAviv",
      "logNumber": 1,
      "name": "OutpostTestName",
      "supplyMeter": 269,
    },
    "meta": Object {
      "feature": "[OUTPOST]",
      "method": "GET",
      "url": "https://jsonplaceholder.typicode.com/posts/1",
    },
    "payload": null,
    "type": "[OUTPOST] API_REQUEST",
  },
  Object {
    "data": Object {
      "id": 0,
      "location": "TelAviv",
      "logNumber": 1,
      "name": "OutpostTestName",
      "supplyMeter": 269,
    },
    "meta": Object {
      "feature": "[OUTPOST]",
    },
    "payload": Object {
      "body": "quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto",
      "id": 1,
      "title": "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
      "userId": 1,
    },
    "type": "[OUTPOST] API_SUCCESS",
  },
  Object {
    "meta": Object {
      "feature": "[LOGS-OFFLINE]",
      "offline": Object {
        "commit": Object {
          "meta": Object {
            "feature": "[LOGS-OFFLINE]",
            "logEntity": Object {
              "calculated": "sunt aut facere repellat provi",
              "content": "Added new outpost id = 0, name = OutpostTestName",
              "id": 0,
            },
          },
          "type": "[LOGS-OFFLINE] ADD_OFFLINE_COMMIT",
        },
        "effect": Object {
          "body": null,
          "method": "GET",
          "url": "https://jsonplaceholder.typicode.com/posts/1",
        },
        "rollback": Object {
          "meta": Object {
            "feature": "[LOGS-OFFLINE]",
            "logEntity": Object {
              "calculated": "sunt aut facere repellat provi",
              "content": "Added new outpost id = 0, name = OutpostTestName",
              "id": 0,
            },
          },
          "type": "[LOGS-OFFLINE] ADD_OFFLINE_ROLLBACK",
        },
      },
    },
    "payload": Object {
      "calculated": "sunt aut facere repellat provi",
      "content": "Added new outpost id = 0, name = OutpostTestName",
      "id": 0,
    },
    "type": "[LOGS-OFFLINE] ADD_OFFLINE",
  },
]
`;

exports[`OutpostAddComponent #008 - API MIDDLEWARE - middleware itself using MOCKs #009 - API MIDDLEWARE CHECK - checking the middleware itself - apiError 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": Object {
          "id": 1,
          "location": "TelAviv",
          "logNumber": 1,
          "name": "OutpostTestName",
          "supplyMeter": 269,
        },
        "meta": Object {
          "feature": "[OUTPOST]",
        },
        "payload": [FetchError: request to https://jsonplaceholder2.typicode.com/posts/a failed, reason: getaddrinfo ENOTFOUND jsonplaceholder2.typicode.com jsonplaceholder2.typicode.com:443],
        "type": "[OUTPOST] API_ERROR",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
